File: MODULE_4/weekly_assignment/customerproduct/Ecom/src/main/java/org/northernarc/ecom/service/ProductService.java
package org.northernarc.ecom.service;

import org.northernarc.ecom.DTO.ProductRequestDTO;
import org.northernarc.ecom.DTO.ProductResponseDTO;
import org.northernarc.ecom.DTO.ProductUpdateDTO;

import java.util.List;

public interface ProductService {

    ProductResponseDTO saveProduct(ProductRequestDTO productDTO);

    List<ProductResponseDTO> getAllProducts();

    ProductResponseDTO getProductById(Integer id);

    ProductResponseDTO updateProduct(Integer id,
                                     ProductUpdateDTO productDTO);

    void deleteProduct(Integer id);

    List<ProductResponseDTO> getProductsByCategory(String category);

    List<ProductResponseDTO> getProductsByBrand(String brand);

    List<ProductResponseDTO> getProductsByName(String name);
}

File: MODULE_6/day5/EcomFrontEndDev/src/app/services/product.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { ProductRequestDTO, ProductResponseDTO, ProductUpdateDTO } from '../dto/product';

@Injectable({
	providedIn: 'root'
})
export class ProductService {
	private readonly baseUrl = 'http://localhost:8080/api/ecom/product';

	constructor(private readonly http: HttpClient) {}

	getAllProducts(): Observable<ProductResponseDTO[]> {
		return this.http.get<ProductResponseDTO[]>(this.baseUrl);
	}

	getProductById(id: number): Observable<ProductResponseDTO> {
		return this.http.get<ProductResponseDTO>(`${this.baseUrl}/${id}`);
	}

	addProduct(payload: ProductRequestDTO): Observable<ProductResponseDTO> {
		return this.http.post<ProductResponseDTO>(this.baseUrl, payload);
	}

	updateProduct(id: number, payload: ProductUpdateDTO): Observable<ProductResponseDTO> {
		return this.http.put<ProductResponseDTO>(`${this.baseUrl}/${id}`, payload);
	}

	deleteProduct(id: number): Observable<void> {
		return this.http.delete<void>(`${this.baseUrl}/${id}`);
	}

	getProductsByCategory(category: string): Observable<ProductResponseDTO[]> {
		return this.http.get<ProductResponseDTO[]>(`${this.baseUrl}/category/${category}`);
	}

	getProductsByBrand(brand: string): Observable<ProductResponseDTO[]> {
		return this.http.get<ProductResponseDTO[]>(`${this.baseUrl}/brand/${brand}`);
	}

	getProductsByName(name: string): Observable<ProductResponseDTO[]> {
		return this.http.get<ProductResponseDTO[]>(`${this.baseUrl}/name/${name}`);
	}
}
